File: libs/ZmqLib/go/rep/responder.go
package responder

import (
	"io/ioutil"

	"github.com/golang/glog"
	zmq "github.com/pebbe/zmq4"
)

// RespondList struct is used to initialize metadata & frame
type RespondList struct {
	MetaData  map[string]string
	FrameBlob string
}

// loadCertificates is used to load all required certs
func loadCertificates(publicKeysDir string, privateKeysDir string) (string, string, string, string) {
	clientPublicKey, err := ioutil.ReadFile(publicKeysDir + "/client.key")
	if err != nil {
		glog.Errorf("Error reading clientPublicKey : %v", err)
		return "", "", "", ""
	}
	clientPublic := string(clientPublicKey)

	clientSecretKey, err := ioutil.ReadFile(privateKeysDir + "/client.key_secret")
	if err != nil {
		glog.Errorf("Error reading clientSecretKey : %v", err)
		return "", "", "", ""
	}
	clientSecret := string(clientSecretKey)

	serverPublicKey, err := ioutil.ReadFile(publicKeysDir + "/server.key")
	if err != nil {
		glog.Errorf("Error reading serverPublicKey : %v", err)
		return "", "", "", ""
	}
	serverPublic := string(serverPublicKey)

	serverSecretKey, err := ioutil.ReadFile(privateKeysDir + "/server.key_secret")
	if err != nil {
		glog.Errorf("Error reading serverSecretKey : %v", err)
		return "", "", "", ""
	}
	serverSecret := string(serverSecretKey)

	return clientPublic, clientSecret, serverPublic, serverSecret
}

// Init method is used to initialize the socket
func Init(address string, publicKeysDir string, privateKeysDir string) (*zmq.Socket, error) {

	// Responder for dev mode
	if publicKeysDir == "" && privateKeysDir == "" {

		//  Create socket for sending
		responder, err := zmq.NewSocket(zmq.REP)
		if err != nil {
			glog.Errorf("Socket creation failed : %v", err)
			return nil, err
		}
		//  Bind to the socket
		errSoc := responder.Bind(address)
		if errSoc != nil {
			glog.Errorf("Error binding to socket : %v", errSoc)
			return nil, errSoc
		}

		return responder, nil
	}

	// Load required certs
	clientPublic, _, _, serverSecret := loadCertificates(publicKeysDir, privateKeysDir)

	zmq.AuthSetVerbose(true)
	zmq.AuthStart()

	//  Tell authenticator to use this public client key
	zmq.AuthCurveAdd("domain1", clientPublic)

	//  Create socket for sending
	responder, err := zmq.NewSocket(zmq.REP)
	if err != nil {
		glog.Errorf("Socket creation failed : %v", err)
		return nil, err
	}

	responder.ServerAuthCurve("domain1", serverSecret)

	//  Bind to the socket
	errSoc := responder.Bind(address)
	if errSoc != nil {
		glog.Errorf("Error binding to socket : %v", errSoc)
		return nil, errSoc
	}

	return responder, nil
}

// Send method is used to send the meta-data and image
func Send(responder *zmq.Socket, callbck func(string) RespondList) error {

	request, err := responder.Recv(0)
	if err != nil {
		glog.Errorf("Error while receiving request : %v", err)
		return err
	}

	// Call the callback with obtained request
	listOfElements := callbck(request)

	glog.Infof("Sending message %s", listOfElements.MetaData)

	//  Send meta-data and frame
	_, errMsg := responder.SendMessage(listOfElements.MetaData, listOfElements.FrameBlob)
	if errMsg != nil {
		glog.Errorf("Error while sending JSON & buffer : %v", errMsg)
		return errMsg
	}
	return nil
}

// Close method is used to close the socket
func Close(responder *zmq.Socket) {
	responder.Close()
}

File: libs/ZmqLib/py/rep/responder.py
import json
import os
import sys
import zmq
import zmq.auth
from zmq.auth.thread import ThreadAuthenticator


class ZMQ_RESPONDER(object):
    """
    The ZMQ_RESPONDER object to send meta-data
    and image frame
    """
    def __init__(self, address, public_keys_dir, secret_keys_dir):
        """
        Constructor to initialize the socket
        and create context
        """
        if public_keys_dir == "" and secret_keys_dir == "":
            self.dev_mode = True
        else:
            self.dev_mode = False

        if self.dev_mode:
            self.context = zmq.Context()
            self.responder = self.context.socket(zmq.REP)
            self.responder.bind(address)
        else:
            # Check if certs exist
            if not (os.path.exists(public_keys_dir) and
                    os.path.exists(secret_keys_dir)):
                print("Certificates are missing - run generate_certificates.py\
                    script first")
                sys.exit(1)

            # Prepare our context and responder
            self.context = zmq.Context()

            # Start an authenticator for this context.
            self.auth = ThreadAuthenticator(self.context)
            self.auth.start()

            # Tell authenticator to use the certificate in a directory
            self.auth.configure_curve(domain='*', location=public_keys_dir)
            self.responder = self.context.socket(zmq.REP)
            server_secret_file = os.path.join(secret_keys_dir,
                                              "zmq_server.key_secret")
            server_public, server_secret = \
                zmq.auth.load_certificate(server_secret_file)

            # Setting public and secret keys of server
            self.responder.curve_secretkey = server_secret
            self.responder.curve_publickey = server_public
            self.responder.curve_server = True
            self.responder.bind(address)

    def send(self, callback):
        """
        This method is used to send the json meta-data
        and numpy frame

        Parameters :
        callback : The callback to be called once the server
        gets the request from client
        """
        flags = 0

        # Receive the request
        request = self.responder.recv_string(flags)

        # Call the callback using the obtained request
        dataList = callback(request)

        # Print the message to be sent
        print("Sending data %s" % (dataList[0]))

        # Sending JSON meta-data
        self.responder.send_json(dataList[0], flags | zmq.SNDMORE)

        # Sending numpy frame
        self.responder.send(dataList[1], flags)

    def send_multipart(self, callback):
        """
        This method is used to send data using send_multipart API

        Parameters :
        callback : The callback to be called once the server
        gets the request from client
        """
        flags = 0

        # Receive the request
        request = self.responder.recv_string(flags)

        # Call the callback using the obtained request
        dataList = callback(request)

        # Print the message to be sent
        print("Sending data %s" % (dataList[0]))

        # Sending JSON meta-data and numpy frame
        self.responder.send_multipart(dataList)

    def close_socket(self):
        # Close the socket and destroy the context
        self.responder.close()
        self.context.term()
        if not self.dev_mode:
            self.auth.stop()
